### DISCUSSION TITLE
Disable recursive collating of lists and dictionaries

### DISCUSSION THREAD

**Discussion by ghost** [pyg-team/pytorch_geometric#4947] [Q&A]
Hello, it seems like the default behavior when collating `HeteroData` is to recursively collate lists and dictionaries? Is it possible to disable this?

**Comment by rusty1s** [comment]
Currently no - this follows PyTorch collate semantics. What's the use-case of not recursively collating?

**Comment by ghost** [comment]
The use-case here is that we have a generic ML pipeline that acts upon both pyg models and other (non-pyg) models. When featurizing the inputs for all model types, in addition to the features we also store some metadata that is required downstream. This metadata is just some nested dicts/lists, but pyg tries to collate this by default -- from testing and profiling, this is actually a significant performance bottleneck for us. 

Some follow-up questions I have are:
1. Is it possible to exclude certain fields from being collated?
2. Is it possible to write a custom collate function / extend the default collate function? (We actually quite like the pyg default collate function for what it does for graph data).

**Reply by rusty1s** [reply]
I see. We can try to get something in like this if you are interested. By default, I think we only collate lists together in case the actually hold PyTorch tensors. This should be quite straightforward to disable via a flag or something similar. WDYT?

**Reply by ghost** [reply]
Yes, it would be great to have an arg that takes the name of attributes that should skipped when collating (including the recursive logic, which is the expensive part for us).

**Reply by rusty1s** [reply]
Do you wanna make a first draft on this?
